File: src/main/java/chat/woowa/woowachat/chat/domain/ChatQueryRepository.java
package chat.woowa.woowachat.chat.domain;


import static chat.woowa.woowachat.chat.domain.QChat.chat;
import static chat.woowa.woowachat.member.domain.QMember.member;
import static org.springframework.util.StringUtils.hasText;

import chat.woowa.woowachat.common.entity.BaseEntity;
import chat.woowa.woowachat.member.domain.Course;
import com.querydsl.core.types.dsl.BooleanExpression;
import com.querydsl.jpa.impl.JPAQuery;
import com.querydsl.jpa.impl.JPAQueryFactory;
import java.util.List;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.springframework.data.support.PageableExecutionUtils;
import org.springframework.stereotype.Repository;

@Repository
public class ChatQueryRepository {

    private final JPAQueryFactory query;

    public ChatQueryRepository(final JPAQueryFactory query) {
        this.query = query;
    }

    public Page<Chat> search(final ChatSearchCond cond, final Pageable pageable) {
        final List<Long> longs = nameAndCourseMatchMemberIds(cond);
        final List<Chat> contents = query.selectFrom(chat)
                .where(
                        titleLikeIgnoreCase(cond.title),
                        chat.memberId.in(longs)
                ).orderBy(chat.createdAt.desc())
                .offset(pageable.getOffset())
                .limit(pageable.getPageSize())
                .fetch();

        JPAQuery<Long> countQuery = query.select(chat.count()).from(chat);

        return PageableExecutionUtils.getPage(contents, pageable, countQuery::fetchOne);
    }

    /* 이름과 코스에 해당하는 회원 id 조회 */
    private List<Long> nameAndCourseMatchMemberIds(final ChatSearchCond cond) {
        return query.selectFrom(member)
                .where(
                        nameLikeIgnoreCase(cond.name),
                        courseEquals(cond.course)).fetch()
                .stream()
                .map(BaseEntity::id)
                .toList();
    }

    private BooleanExpression nameLikeIgnoreCase(final String name) {
        return hasText(name) ? member.name.likeIgnoreCase("%" + name.strip() + "%") : null;
    }

    private BooleanExpression courseEquals(final Course course) {
        return (course == null) ? null : member.course.eq(course);
    }

    private BooleanExpression titleLikeIgnoreCase(final String title) {
        return hasText(title) ? chat.title.likeIgnoreCase("%" + title.strip() + "%") : null;
    }

    public record ChatSearchCond(
            String name,
            String title,
            Course course
    ) {
    }
}

File: src/main/kotlin/chat/teco/tecochat/query/ChatQueryRepository.kt
package chat.teco.tecochat.query

import chat.teco.tecochat.domain.chat.Chat
import chat.teco.tecochat.domain.chat.QChat.chat
import chat.teco.tecochat.domain.chatlike.QChatLike.chatLike
import chat.teco.tecochat.domain.member.QMember.member
import chat.teco.tecochat.support.domain.BaseEntity
import chat.teco.tecochat.support.querydsl.orderByNotEmpty
import chat.teco.tecochat.support.querydsl.whereNotEmpty
import com.querydsl.jpa.impl.JPAQueryFactory
import org.springframework.data.domain.Page
import org.springframework.data.domain.Pageable
import org.springframework.data.support.PageableExecutionUtils
import org.springframework.stereotype.Repository

@Repository
class ChatQueryRepository(
    private val query: JPAQueryFactory,
) {
    fun search(cond: ChatSearchCond, pageable: Pageable): Page<Chat> {
        val longs = nameAndCourseMatchMemberIds(cond)
        val contents = query.selectFrom(chat)
            .leftJoin(chatLike)
            .on(chatLike.chatId.eq(chat.id))
            .distinct()
            .whereNotEmpty(cond.likeCond?.dataCondition()) { chatLike.createdAt.goe(it) }
            .whereNotEmpty(cond.title) { chat.title.likeIgnoreCase("%" + it.trim() + "%") }
            .where(chat.memberId.`in`(longs))
            .orderByNotEmpty(cond.likeCond) { chat.likeCount.desc() }
            .orderBy(chat.createdAt.desc())
            .offset(pageable.offset)
            .limit(pageable.pageSize.toLong())
            .fetch()
        val countQuery = query.select(chat.count()).from(chat)
        return PageableExecutionUtils.getPage(contents, pageable) { countQuery.fetchOne()!! }
    }

    /* 이름과 코스에 해당하는 회원 id 조회 */
    private fun nameAndCourseMatchMemberIds(cond: ChatSearchCond): List<Long> {
        return query.selectFrom(member)
            .whereNotEmpty(cond.name) { member.name.likeIgnoreCase("%" + it.trim() + "%") }
            .whereNotEmpty(cond.course) { member.course.eq(it) }
            .fetch()
            .map(BaseEntity::id)
    }
}

